### PR TITLE
[BEAM-4742] allow custom docker image in portable runner

### PR DESCRIPTION
**Pull request by ryan-williams**
Allow specifying a docker image for the portable runner to use, via `DOCKER_IMAGE` env var

Also: make output directory in wordcount example, if it doesn't exist.

R: @angoenka 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

File: sdks/python/apache_beam/examples/wordcount.py [hunk 1]
```diff
@@ -111,6 +113,10 @@ def format_result(word_count):
 
   output = counts | 'format' >> beam.Map(format_result)
 
+  out_dir = os.path.dirname(known_args.output)
+  if not FileSystems.exists(out_dir):
```
Comment (lukecwik) [inline sdks/python/apache_beam/examples/wordcount.py]: I believe the expectation should be that any output path should be created during pipeline execution and not by the driver program creating the pipeline.

Please revert this change to wordcount and fix the filesystem implementation to create any necessary directories instead.

File: sdks/python/apache_beam/runners/portability/portable_runner.py [hunk 2]
```diff
@@ -59,7 +59,15 @@ def __init__(self, is_embedded_fnapi_runner=False):
 
   @staticmethod
   def default_docker_image():
-    if 'USER' in os.environ:
+    if 'DOCKER_IMAGE' in os.environ:
```
Comment (lukecwik) [inline sdks/python/apache_beam/runners/portability/portable_runner.py]: This is already controlled by the flag `--harness_docker_image`:
https://github.com/apache/beam/blob/385faa713951813371dffaf654b5dc8d96e27aa1/sdks/python/apache_beam/options/pipeline_options.py#L648

Do you still want to make the default container selection be based off of `DOCKER_IMAGE`?
If yes, should it specify the full path and not assume the user wants the `:latest` suffix?

File: sdks/python/apache_beam/examples/wordcount.py (hunk 1 above)
Comment (ryan-williams) [reply sdks/python/apache_beam/examples/wordcount.py]: interesting, I originally [made a change to `LocalFileSystem` to create directories on `open`](https://github.com/ryan-williams/beam/commit/25868025c2ead0b695d0dde46b6d4e3d19a4923a), but I wasn't sure if that was the right semantics; it sounds like you're saying it is?

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 2 above)
Comment (ryan-williams) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: ah, yea, I just saw the pipeline option for this as well! thanks for pointing it out.

Comment (ryan-williams) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: (I'll revert this part of the change, I don't think it's necessary)

File: sdks/python/apache_beam/examples/wordcount.py (hunk 1 above)
Comment (ryan-williams) [reply sdks/python/apache_beam/examples/wordcount.py]: I filed https://github.com/apache/beam/pull/5903 with that change; can close this in favor of that if that's what you prefer, thanks!

Comment (lukecwik) [reply sdks/python/apache_beam/examples/wordcount.py]: Thanks, taking a look at #5903.

**Comment by ryan-williams** [comment]
closing in favor of #5903, thanks @lukecwik
